pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -24,130 +24,183 @@ import (
 	"google.golang.org/grpc"
 	grpccodes "google.golang.org/grpc/codes"
 	grpcstatus "google.golang.org/grpc/status"
-	"k8s.io/klog/v2"
 
+	"k8s.io/klog/v2"
 	drapbv1alpha2 "k8s.io/kubelet/pkg/apis/dra/v1alpha2"
 	drapb "k8s.io/kubelet/pkg/apis/dra/v1alpha3"
 )
 
 const PluginClientTimeout = 45 * time.Second
 
-// draPluginClient encapsulates all dra plugin methods.
-type draPluginClient struct {
-	pluginName string
-	plugin     *Plugin
-}
+type (
+	nodePrepareResourceHandler   func(context.Context, *grpc.ClientConn, *plugin, *drapb.NodePrepareResourcesRequest) (*drapb.NodePrepareResourcesResponse, error)
+	nodeUnprepareResourceHandler func(context.Context, *grpc.ClientConn, *plugin, *drapb.NodeUnprepareResourcesRequest) (*drapb.NodeUnprepareResourcesResponse, error)
+)
+
+var (
+	nodePrepareResourceHandlers = map[string]nodePrepareResourceHandler{
+		v1alpha2Version: v1alpha2NodePrepareResouce,
+		v1alpha3Version: v1alpha3NodePrepareResouce,
+	}
+
+	nodeUnprepareResourceHandlers = map[string]nodeUnprepareResourceHandler{
+		v1alpha2Version: v1alpha2NodeUnprepareResouce,
+		v1alpha3Version: v1alpha3NodeUnprepareResouce,
+	}
+)
 
 func NewDRAPluginClient(pluginName string) (drapb.NodeClient, error) {
 	if pluginName == "" {
 		return nil, fmt.Errorf("plugin name is empty")
 	}
 
-	existingPlugin := draPlugins.Get(pluginName)
+	existingPlugin := draPlugins.get(pluginName)
 	if existingPlugin == nil {
 		return nil, fmt.Errorf("plugin name %s not found in the list of registered DRA plugins", pluginName)
 	}
 
-	return &draPluginClient{
-		pluginName: pluginName,
-		plugin:     existingPlugin,
-	}, nil
+	return existingPlugin, nil
 }
 
-func (r *draPluginClient) NodePrepareResources(
+func (p *plugin) NodePrepareResources(
 	ctx context.Context,
 	req *drapb.NodePrepareResourcesRequest,
 	opts ...grpc.CallOption,
-) (resp *drapb.NodePrepareResourcesResponse, err error) {
+) (*drapb.NodePrepareResourcesResponse, error) {
 	logger := klog.FromContext(ctx)
 	logger.V(4).Info(log("calling NodePrepareResources rpc"), "request", req)
-	defer logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", resp, "err", err)
```

Comment:
I guess by removing this defer we lose the ability to log this when there is an error condition?

Reply:
I have updated the code to catch all error condition in 0053f2a7be8